### PR TITLE
fix flow in finally blocks

### LINKED ISSUE #11665: In finally block, local variable is incorrectly typed as never.
**TypeScript Version:**  2.0.3

**Code**

With `--strictNullChecks`,

```ts
class A {
  constructor() { }
  method() { }
}

let a: A | null = null;

try {
  a = new A();
} finally {
  if (a) {
    a.method();
  }
}
```

**Expected behavior:**

I expect this to compile. TypeScript should be able to deduce that `a` can be of type `A` at the `a.method()` call site.

**Actual behavior:** 

I get an error message

`Property 'method' does not exist on type 'never'.`

Note that it works fine if the try and finally lines are removed.

### PR DESCRIPTION
**Pull request by vladima**
fixes #11665.

Make initial flow in finally block a combination from start flow in `try`, final flow in `try` and final flow in `catch` blocks

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -984,24 +984,44 @@ namespace ts {
...
                 bind(node.finallyBlock);
+                // if flow after finally is unreachable - keep it
+                // otherwise check if flows after try and after catch are unreachable 
+                // if yes - convert current flow to unreachable
+                // i.e.
+                // try { return "1" } finally { console.log(1); }
+                // console.log(2); // this line should be unreachable even if flow falls out of finally block
+                if (!(currentFlow.flags & FlowFlags.Unreachable)) {
+                    if ((flowAfterTry.flags & FlowFlags.Unreachable) && (flowAfterCatch.flags & FlowFlags.Unreachable)) {
+                        currentFlow = flowAfterTry == reportedUnreachableFlow || flowAfterCatch === reportedUnreachableFlow
```
Comment (sandersn) [inline src/compiler/binder.ts]: typo: `==` not `===`

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -984,24 +984,44 @@ namespace ts {
...
                 bind(node.catchClause);
-                addAntecedent(postFinallyLabel, currentFlow);
+                addAntecedent(preFinallyLabel, currentFlow);
+
+                flowAfterCatch = currentFlow;
             }
             if (node.finallyBlock) {
-                currentFlow = preTryFlow;
+                // in finally flow is combined from pre-try/flow from try/flow from catch
+                // pre-flow is necessary to make sure that finally is reachable even if finaly flows in both try and finally blocks are unreachable
```
Comment (sandersn) [inline src/compiler/binder.ts]: typo:finaly

Comment (vladima) [inline src/compiler/binder.ts]: ✔️

File: src/compiler/binder.ts (hunk 1 above)
Comment (vladima) [inline src/compiler/binder.ts]: ✔️

**Pull request merged**